commit 65cd21b058bff6371ae315f886c824524fade82a
Author: Wu Zheng <wu.zheng@intel.com>
Date:   Sun Sep 29 16:25:12 2013 +0800

    TIZEN DBUS configure file

diff --git a/obexd/src/obex.service.in b/obexd/src/obex.service.in
index bca3aef..a44bb89 100644
--- a/obexd/src/obex.service.in
+++ b/obexd/src/obex.service.in
@@ -4,7 +4,7 @@ Description=Bluetooth OBEX service
 [Service]
 Type=dbus
 BusName=org.bluez.obex
-ExecStart=@libexecdir@/obexd
+ExecStart=/bin/sh -c 'if [ -z `ps ax | grep -v grep | grep obexd` ] ; then exec @libexecdir@/obexd -d --noplugin=syncevolution,pcsuite,irmc --symlinks -r /opt/usr/media; fi'
 
 [Install]
 Alias=dbus-org.bluez.obex.service
